### PR TITLE
Add Matthews correlation coefficient to metrics

### PR DESCRIPTION
**Pull request by kuza55**
I needed this for a Kaggle competition and it seemed useful in general so I thought I'd contribute it back.

Let me know if there are changes you'd like before merging it in.

### CODE REVIEW

**Comment by ternaus** [comment]
It looks like there are couple issues with this metric:

1. Argument of the sqrt function may become small, and the ratio will explode. A probable solution is to apply -log(x) to this metric.
2. Argument of the sqrt function may become zero => Possible solution to add smoothing coefficient. like A/B => (A+1)/(B+1), similar to what people did in Kaggle Ultrasound competition, where they used Dice coefficient as metric.

**Comment by kuza55** [comment]
@ternaus 

So, I'm not keen on mucking with the formulation since then we'd have some random metric that no-one has evaluated.

You're right that this can be NaN if either all your predictions or all of your ground truth are from a single class, which sort of seems fine? Either case seems sort of pathological.

I don't think you're right that the ratio can explode though since all the components of the sqrt are sums of integers, not ratios, so besides the pathological case you pointed out, the smallest value it can go to is 1.

File: keras/metrics.py [hunk 1]
```diff
@@ -71,6 +71,25 @@ def cosine_proximity(y_true, y_pred):
     return -K.mean(y_true * y_pred)
 
 
+def matthews(y_true, y_pred):
```
Comment (fchollet) [inline keras/metrics.py]: Should be renamed to `matthews_correlation` or something similar.

File: keras/metrics.py [hunk 2]
```diff
@@ -71,6 +71,25 @@ def cosine_proximity(y_true, y_pred):
...
+    ''' Matthews correlation coefficient
+    '''
+    y_pred_pos = K.round(y_pred)
+    y_pred_neg = 1 - y_pred_pos
```
Comment (fchollet) [inline keras/metrics.py]: The targets / predictions are assumed to be in the [0, 1] interval but that is not enforced. It should be.

File: keras/metrics.py [hunk 3]
```diff
@@ -71,6 +71,25 @@ def cosine_proximity(y_true, y_pred):
...
+    
+    y_pos = y_true
+    y_neg = 1 - y_true
+
+    TP = K.sum(y_pos * y_pred_pos)
+    TN = K.sum(y_neg * y_pred_neg)
```
Comment (fchollet) [inline keras/metrics.py]: Capital variable names is not appropriate for scalar variables

File: keras/metrics.py [hunk 4]
```diff
@@ -71,6 +71,25 @@ def cosine_proximity(y_true, y_pred):
...
+
+    FP = K.sum(1 - y_neg * y_pred_pos)
+    FN = K.sum(1 - y_pos * y_pred_neg)
+
+    MCC = (TP*TN-FP*FN)/K.sqrt((TP+FP)*(TP+FN)*(TN+FP)*(TN+FN))
```
Comment (fchollet) [inline keras/metrics.py]: This should be fuzzed to avoid NaNs

File: keras/metrics.py (hunk 2 above)
Comment (kuza55) [inline keras/metrics.py]: Do you have any ideas on how to enforce this? IIRC y_true and y_pred are tensors, so there's no trivial way of turning these into values, and doing so would hurt performance when this is a training metric.

I could probably come up with some tensor operations to make this get done on the GPU, but it's not fully clear how I would ensure they get validated; in tensorflow it might be possible to hack some hideous assert/group op thing together, but it seems overkill for a simple metric; I have no idea about Theano.

File: keras/metrics.py (hunk 4 above)
Comment (kuza55) [inline keras/metrics.py]: So, this can be NaN if either all your predictions or all of your ground truth are from a single class, though that is there by definition of the metric, is that an issue? Could you give me some pointers to the sort of fuzzing you want done? I'm not super familiar with numerical issues.

File: keras/metrics.py (hunk 2 above)
Comment (fchollet) [inline keras/metrics.py]: You could just clip to [0, 1].

File: keras/metrics.py (hunk 4 above)
Comment (fchollet) [inline keras/metrics.py]: `+ K.epsilon()` in the denominator.

File: keras/metrics.py [hunk 5]
```diff
@@ -71,6 +71,23 @@ def cosine_proximity(y_true, y_pred):
     return -K.mean(y_true * y_pred)
 
 
+def matthews_corrcoef(y_true, y_pred):
+    ''' Matthews correlation coefficient
+    '''
+    y_pred_pos = K.round(K.clip(y_pred, 0 ,1))
```
Comment (fchollet) [inline keras/metrics.py]: PEP8

File: keras/metrics.py [hunk 6]
```diff
@@ -71,6 +71,23 @@ def cosine_proximity(y_true, y_pred):
...
+    y_pos = K.round(K.clip(y_true, 0, 1))
+    y_neg = 1 - y_pos
+
+    tp = K.sum(y_pos * y_pred_pos)
+    tn = K.sum(y_neg * y_pred_neg)
+
+    fp = K.sum(1 - y_neg * y_pred_pos)
+    fn = K.sum(1 - y_pos * y_pred_neg)
+    
+    return (tp*tn-fp*fn)/(K.sqrt((tp+fp)*(tp+fn)*(tn+fp)*(tn+fn))+K.epsilon())
```
Comment (fchollet) [inline keras/metrics.py]: PEP8

File: keras/metrics.py [hunk 7]
```diff
@@ -71,6 +71,23 @@ def cosine_proximity(y_true, y_pred):
     return -K.mean(y_true * y_pred)
 
 
+def matthews_corrcoef(y_true, y_pred):
```
Comment (fchollet) [inline keras/metrics.py]: `mathews_correlation` or `matthews_correlation_coefficient`

File: keras/metrics.py [hunk 8]
```diff
@@ -71,6 +71,23 @@ def cosine_proximity(y_true, y_pred):
...
+    y_pos = K.round(K.clip(y_true, 0, 1))
+    y_neg = 1 - y_pos
+
+    tp = K.sum(y_pos * y_pred_pos)
+    tn = K.sum(y_neg * y_pred_neg)
+
+    fp = K.sum(1 - y_neg * y_pred_pos)
+    fn = K.sum(1 - y_pos * y_pred_neg)
+
+    return (tp*tn-fp*fn)/(K.sqrt((tp+fp)*(tp+fn)*(tn+fp)*(tn+fn))+K.epsilon())
```
Comment (fchollet) [inline keras/metrics.py]: PEP8 specifies that there should be spaces around operators. Also this is a very long line, so you might want to break it up into two lines.

File: keras/metrics.py [hunk 9]
```diff
@@ -71,6 +71,26 @@ def cosine_proximity(y_true, y_pred):
     return -K.mean(y_true * y_pred)
 
 
+def mathews_correlation(y_true, y_pred):
```
Comment (fchollet) [inline keras/metrics.py]: with 2 t's

Comment (kuza55) [inline keras/metrics.py]: Hah, I copied your typo :)
